pkg/envoy/envoy.go

```diff
@@ -134,139 +135,141 @@ func GetEnvoyVersion() string {
-				// Envoy process crash - after closing that channel by the goroutime,
-				// the loop continues, the channel is recreated and the new process
-				// is watched again.
-				crashCh := make(chan struct{})
-				go func() {
-					if err := cmd.Wait(); err != nil {
-						log.WithError(err).Warn("Envoy: Proxy crashed")
-						// Avoid busy loop & hogging CPU resources by waiting before restarting envoy.
-						time.Sleep(100 * time.Millisecond)
-					}
-					close(crashCh)
-				}()
+			log.Infof("Envoy: Proxy started with pid %d", cmd.Process.Pid)
+			metrics.SubprocessStart.WithLabelValues(ciliumEnvoy).Inc()
+
+			// We do not return after a successful start, but watch the Envoy process
+			// and restart it if it crashes.
+			// Waiting for the process execution is done in the goroutime.
+			// The purpose of the "crash channel" is to inform the loop about their
+			// Envoy process crash - after closing that channel by the goroutime,
+			// the loop continues, the channel is recreated and the new process
+			// is watched again.
+			crashCh := make(chan struct{})
+			go func() {
+				if err := cmd.Wait(); err != nil {
+					log.WithError(err).Warn("Envoy: Proxy crashed")
+					// Avoid busy loop & hogging CPU resources by waiting before restarting envoy.
+					time.Sleep(100 * time.Millisecond)
+				}
+				close(crashCh)
+			}()
 
-				select {
-				case <-crashCh:
-					// Start Envoy again
-					continue
-				case <-e.stopCh:
-					log.Infof("Envoy: Stopping proxy with pid %d", cmd.Process.Pid)
-					if err := e.admin.quit(); err != nil {
-						log.WithError(err).Fatalf("Envoy: Envoy admin quit failed, killing process with pid %d", cmd.Process.Pid)
-
-						if err := cmd.Process.Kill(); err != nil {
-							log.WithError(err).Fatal("Envoy: Stopping Envoy failed")
-							e.errCh <- err
-						}
+			select {
+			case <-crashCh:
+				// Start Envoy again
+				continue
+			case <-e.stopCh:
+				log.Infof("Envoy: Stopping proxy with pid %d", cmd.Process.Pid)
+				if err := e.admin.quit(); err != nil {
+					log.WithError(err).Fatalf("Envoy: Envoy admin quit failed, killing process with pid %d", cmd.Process.Pid)
+
+					if err := cmd.Process.Kill(); err != nil {
+						log.WithError(err).Fatal("Envoy: Stopping Envoy failed")
+						e.errCh <- err
 					}
-					close(e.errCh)
-					return
 				}
+				close(e.errCh)
+				return
 			}
-		}()
-
-		if <-started {
-			return e
 		}
+	}()
 
-		return nil
-	} else {
-		// Case Envoy as DaemonSet
+	if <-started {
 		return e
 	}
+
+	return nil
+}
+
+func NewEnvoyAdminClient(envoySocketDir string) *EnvoyAdminClient {
+	// Have to use a fake IP address:port even when we Dial to a Unix domain socket.
+	// The address:port will be visible to Envoy as ':authority', but its value is
+	// not meaningful.
+	// Not using the normal localhost address to make it obvious that we are not
+	// connecting to Envoy's admin interface via the IP stack.
+	adminAddress := "192.0.2.34:56"
```

Comment:
can you use some host name/alias like envoy-admin here ?

Reply:
@sayboras 

yes it is possible. got rid of it (even though the author wrote it this way on purpose) - but let the necessary parts of the comment there.